### PR TITLE
WIP: Data class Rethink

### PR DESCRIPTION
**Pull request by andersbogsnes**
Separating out Data from the Model in order to allow for ease of use when changing datasources for the model

- [x] closes issue #150 
- [x] Tests added
- [x] Passes flake8
- [ ] Updated CHANGELOG
- [ ] Updated README.md

### CODE REVIEW

**Comment by andersbogsnes** [comment]
@thomasfrederikhoeck Foreløbig WIP

File: dev-requirements.in [hunk 1]
```diff
@@ -0,0 +1,4 @@
+-rrequirements.txt
```
Comment (thomasfrederikhoeck) [inline dev-requirements.in]: Does it need a space here?

File: src/ml_tooling/__init__.py [hunk 2]
```diff
@@ -1,6 +1,16 @@
-# flake8: noqa
 from .baseclass import ModelData
-import pkg_resources
 
-__version__ = pkg_resources.get_distribution("ml_tooling").version
-__all__ = ["ModelData", "__version__"]
+__all__ = ["ModelData"]
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/__init__.py]: Maybe include the new data class here, as it would always be used.

File: src/ml_tooling/baseclass.py [hunk 3]
```diff
@@ -327,7 +233,12 @@ def save_estimator(
...
-        self, input_data: Any, proba: bool = False, use_index: bool = False
+        self,
+        data: DataSet,
+        *args,
+        proba: bool = False,
+        use_index: bool = False,
+        **kwargs,
     ) -> pd.DataFrame:
         """Makes a prediction given an input. For example a customer number.
         Passed to the implemented :meth:`get_prediction_data` method and calls `predict()`
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: Not quite accurate. I guess it should be something with the the `DataSet` exposing `get_prediction_data`

File: src/ml_tooling/baseclass.py [hunk 4]
```diff
@@ -471,15 +382,15 @@ def train_estimator(self) -> "ModelData":
...
         self.result = None
         logger.info("Estimator trained!")
 
         return self
 
     def score_estimator(
-        self, metric: Optional[str] = None, cv: Optional[int] = False
+        self, data: DataSet, metric: Optional[str] = None, cv: Optional[int] = False
     ) -> "Result":
         """Loads all training data and trains the estimator on it, using a train_test split.
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: Train test split is done in DataSet, so it should just say it trains on the train set from Dataset and socres on the test set from Dataset

File: src/ml_tooling/baseclass.py [hunk 5]
```diff
@@ -582,7 +504,7 @@ def gridsearch(
                 parameters=parameters,
             )
             for parameters, (train, test) in product(
-                param_grid, cv.split(train_x, train_y, None)
+                param_grid, cv.split(data.train_x, data.train_y, None)
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: Maybe the generation of splits should be de handled by a method in dataset. That way we can easliy define it for times series and grouped observations.

File: src/ml_tooling/baseclass.py [hunk 6]
```diff
@@ -471,15 +382,15 @@ def train_estimator(self) -> "ModelData":
...
+        self.estimator.fit(data.x, data.y)
         # Prevent confusion, as train_estimator does not return a result
         self.result = None
         logger.info("Estimator trained!")
 
         return self
 
     def score_estimator(
-        self, metric: Optional[str] = None, cv: Optional[int] = False
+        self, data: DataSet, metric: Optional[str] = None, cv: Optional[int] = False
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: We should also note that score estimator uses only train for cv and train+test if non-cv

File: src/ml_tooling/baseclass.py [hunk 7]
```diff
@@ -643,12 +565,14 @@ def log(self, run_name: str):
...
 
-    def _score_estimator(self, estimator, metric: str) -> Result:
+    def _score_estimator(self, data, estimator, metric: str) -> Result:
         """
         Scores estimator with a given score function.
 
         Parameters
         ----------
+        data: DataSet
+            An instantiated DataSet object
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: with train test split

File: src/ml_tooling/data/base_data.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            Whether or not to shuffle the data
+        :param test_size:
+            What percentage of the data will be part of the test set
+         :param seed:
+            Random seed for train_test_split
+        :return:
+            self
+        """
+
+        self.train_x, self.test_x, self.train_y, self.test_y = train_test_split(
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/data/base_data.py]: Here we assume that we always want to use a simple train_test_split. Maybe we should create a wrapper for a Shuffler where the default used Shuffler is ShuffleSplit. That way one could also easliy use GroupShuffleSplit, TimeSeriesSplit......

File: src/ml_tooling/baseclass.py (hunk 5 above)
Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: Agreed, but think that's a more advanced feature we would have to think about how to implement - Make this an issue?

File: src/ml_tooling/utils.py [hunk 9]
```diff
@@ -69,7 +74,7 @@ def find_estimator_file(path: str) -> pathlib.Path:
     return newest_match
 
 
-def _get_estimator_name(clf) -> str:
+def _get_estimator_name(clf: BaseEstimator) -> str:
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/utils.py]: This assumes that we would always use a Sklearn based estimator. Isn't that to much to assume?

File: src/ml_tooling/data/sql.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import abc
+from typing import Tuple
+
+from ml_tooling.data.base_data import DataSet
+from ml_tooling.utils import DataType
+import sqlalchemy as sa
+from sqlalchemy.engine import Connectable
+
+
+class SQLDataSet(DataSet, metaclass=abc.ABCMeta):
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/data/sql.py]: I would be convenient to have dump method too. That way the data could be dumped to another DB accessibile in porduction.

File: tests/test_data.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import pytest
+from ml_tooling.utils import DataSetError
+
+
+def test_sqldataset_works_correctly(test_sqldata):
+    assert test_sqldata._x is None
+    features = test_sqldata.x
+    assert test_sqldata._x is not None
```
Comment (thomasfrederikhoeck) [inline tests/test_data.py]: Same test twice? Maybe test _y?

File: src/ml_tooling/__init__.py (hunk 2 above)
Comment (andersbogsnes) [reply src/ml_tooling/__init__.py]: There are three right now, Dataset, FileDataset and SQLDataset - I like leaving them under their namespace, so it's easier to discover what types there are

File: src/ml_tooling/data/base_data.py (hunk 8 above)
Comment (andersbogsnes) [reply src/ml_tooling/data/base_data.py]: I would probably do that as part of the implementation of internal cross-validation splitting - then this would be calling that with CV=1

File: src/ml_tooling/baseclass.py (hunk 5 above)
Comment (thomasfrederikhoeck) [reply src/ml_tooling/baseclass.py]: Sure :-) I will create an issue.

File: src/ml_tooling/utils.py (hunk 9 above)
Comment (andersbogsnes) [reply src/ml_tooling/utils.py]: It's a typehint, so it doesn't change the API, but so far we're depending on Sklearn API - our custom models would also inherit from BaseEstimator. 

If we want to switch to another framework, we would have to change the entire API anyways

File: src/ml_tooling/data/sql.py (hunk 10 above)
Comment (andersbogsnes) [reply src/ml_tooling/data/sql.py]: I think we need some more experience with how that would look - we can do some adhocs when creating our subclass in our current APIs and add it in to ML tooling when we have something that works?

File: src/ml_tooling/data/base_data.py (hunk 8 above)
Comment (thomasfrederikhoeck) [reply src/ml_tooling/data/base_data.py]: I have created an issue we can work on later then.

File: tests/test_visualize.py [hunk 12]
```diff
@@ -272,7 +257,7 @@ def test_lift_curve_have_correct_data(self, classifier):
         assert "Lift" == ax.get_ylabel()
         assert "% of Data" == ax.get_xlabel()
         assert pytest.approx(19.5) == np.sum(ax.lines[0].get_xdata())
-        assert pytest.approx(49.849, rel=0.0001) == np.sum(ax.lines[0].get_ydata())
+        assert pytest.approx(60.291, rel=0.0001) == np.sum(ax.lines[0].get_ydata())
```
Comment (thomasfrederikhoeck) [inline tests/test_visualize.py]: Shuffled data?

File: tests/test_data.py (hunk 11 above)
Comment (andersbogsnes) [reply tests/test_data.py]: It's testing the property works - `._x` should start out as None, but if `.x` is accessed, it should populate `._x` with data, i.e not be None

File: tests/test_visualize.py (hunk 12 above)
Comment (andersbogsnes) [reply tests/test_visualize.py]: Yeah, the dataset gets moved around a bit, so the randomness occurs in different order

File: src/ml_tooling/baseclass.py [hunk 13]
```diff
@@ -367,7 +291,7 @@ def test_estimators(
 
         return cls(best_estimator), ResultGroup(results)
 
-    def train_estimator(self, data: Dataset) -> "ModelData":
+    def train_estimator(self, data: Dataset) -> "Model":
```
Comment (thomasfrederikhoeck) [inline src/ml_tooling/baseclass.py]: Why is the typehint in quotes?

Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: It's referring to itself, which is not defined yet, so quotes are valid for typehinting as well, in this case, but also if you can't import due to circular imports or something similar
